engines/order_management/lib/order_management/reports/enterprise_fee_summary/scope.rb

```diff
@@ -0,0 +1,361 @@
+          filter_scope("spree_orders.completed_at <= ?", params.end_at) if params.end_at.present?
+        end
+
+        def filter_by_distribution(params)
+          filter_scope(spree_orders: { distributor_id: params.distributor_ids }) \
+            if params.distributor_ids.present?
+          filter_scope(spree_products: { supplier_id: params.producer_ids }) \
+            if params.producer_ids.present?
+          filter_scope(spree_orders: { order_cycle_id: params.order_cycle_ids }) \
+            if params.order_cycle_ids.present?
+        end
+
+        def filter_by_fee(params)
+          filter_scope(enterprise_fees: { id: params.enterprise_fee_ids }) \
+            if params.enterprise_fee_ids.present?
+          filter_scope(spree_shipping_methods: { id: params.shipping_method_ids }) \
+            if params.shipping_method_ids.present?
+          filter_scope(spree_payment_methods: { id: params.payment_method_ids }) \
+            if params.payment_method_ids.present?
+        end
+
+        def group_data
+          chain_to_scope do
+            group("enterprise_fees.id", "enterprises.id", "customers.id", "hubs.id",
+                  "spree_payment_methods.id", "spree_shipping_methods.id",
+                  "adjustment_metadata.enterprise_role", "spree_tax_categories.id",
+                  "product_tax_categories.id", "incoming_exchange_enterprises.id",
+                  "outgoing_exchange_enterprises.id")
+          end
+        end
+
+        def select_attributes
+          chain_to_scope do
+            select(
+              <<-JOIN_STRING.strip_heredoc
+                SUM(spree_adjustments.amount) AS total_amount, spree_payment_methods.name AS
+                  payment_method_name, spree_shipping_methods.name AS shipping_method_name,
+                  hubs.name AS hub_name, enterprises.name AS enterprise_name,
+                  enterprise_fees.fee_type AS fee_type, customers.name AS customer_name,
+                  customers.email AS customer_email, enterprise_fees.fee_type AS fee_type,
+                  enterprise_fees.name AS fee_name, spree_tax_categories.name AS tax_category_name,
+                  product_tax_categories.name AS product_tax_category_name,
+                  adjustment_metadata.enterprise_role AS placement_enterprise_role,
+                  incoming_exchange_enterprises.name AS incoming_exchange_enterprise_name,
+                  outgoing_exchange_enterprises.name AS outgoing_exchange_enterprise_name
+              JOIN_STRING
+            )
+          end
+        end
+
+        def chain_to_scope(&block)
+          @scope = @scope.instance_eval(&block)
+          self
+        end
+
+        def join_scope(join_string)
+          chain_to_scope do
+            joins(join_string)
+          end
+        end
+
+        def filter_scope(*args)
+          chain_to_scope do
+            where(*args)
+          end
+        end
+      end
+    end
+  end
+end
```

Comment:
:metal: SQL! I think though that by making use AR's DSL we can make this a bit more maintainable. On the other hand, I feel like we just put together all these queries into a class because the just touch the same tables, leading to poor cohesion. It :nose: too attached to AR scopes.

When I have to implement a rather complex query I find the query object pattern very handy. I first read about it in the famous Code Climate's post https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/ and the folks at Thoughtbot have a nice practical case as well in https://robots.thoughtbot.com/active-record-eager-loading-with-query-objects-and-decorators#the-query-object

Reply:
nice! I studied this pattern in my MsC back in 2004 when "P of EAA" was a new book!!!
https://martinfowler.com/eaaCatalog/queryObject.html
In this link you find a more advanced case where the query object can be composed of variations/subclasses/query_criteria, I dont think we need this here but I am happy to share some history :-D

Reply:
I did not use named scopes here because Rails 3 doesn't allow use of named scopes for left joins, many of the joins need custom join names (for example, `enterprises` and `exchanges` tables are used more than once), and there are many `ON` conditions that presume certain joins are already available. These makes it complicated to ensure that scope chaining would still work with other existing named scopes.

I will see if there is a reliable gem for Rails 3 that would allow us to use left join for named scopes, and if yes, see what is reusable enough and can be moved to a model. :slightly_smiling_face:

Reply:
Thank you for the resources, @sauloperez and @luisramos0! :slightly_smiling_face: So much to learn, and motivated to study more so I can share some back too.

But for now, sharing that I was mind-blown :heart_eyes: when I realized how the very simple the concept of returning `self` is how `ActiveRecord::Relation` allows method chaining:

```ruby
def chain_to_scope(&block)
  @scope = @scope.instance_eval(&block)
  self
end
```

Hehe, it might be obvious, but I didn't really think about it before. Used method chaining in an initial version of the Scope class.

Reply:
yeah, chaining is commonly done by returning self ("return this" in java and javascript). And hopefully it's a form of the builder pattern, otherwise it's probably just someone trying to be smart :-)

Reply:
my point was more about having bits and pieces of AR where we can so things are a bit more maintainable. Of course there's no support for LEFT JOIN but we can have just these as SQL or [Arel](https://jpospisil.com/2014/06/16/the-definitive-guide-to-arel-the-sql-manager-for-ruby.html) if it makes sense.

Reply:
Thanks @sauloperez! Okay, I will revisit. I've never used `Arel` directly yet.

Reply:
I will create a tech debt issue for investigating reusing some model associations here or using Arel. I'm marking this review thread resolved already. :slightly_smiling_face: But if you have objections, feel free to undo.